### ISSUE TITLE
Strategy advice on modified base detection approach using the remora API

### ISSUE THREAD

**Issue by drivenbyentropy** [nanoporetech/remora#110]
Let me begin by thanking you for providing us with the great API that was implemented in the later versions of remora. It feels like a big step forward in terms of providing consistent and easy access to signal based read analysis. 

I am currently aiming at levering the API to implement a pipeline for non-canonical base detection using a sample compare approach similar to the one used in Tombo and I have a conceptual question that I was hoping you could shed some light on. It mainly stems from my confusion regarding the difference between basecall-anchored metrics and reference-anchored metrics and how insertions and deletions are handled in the latter. 

In order to build my canonical model for each nucleotide position of a reference genome, I need to assign the specific base of all reads covering the genomic position to that locus. I see two possible approaches here:

1) Make use of the ```RefRegion``` class in conjunction with ```io.get_reads_reference_regions()``` and to extract all reads which cover a loci of interest and process the signal from there. 

2) Use ```io.get_reg_bam_reads()``` in conjunction with ```io.get_io_reads()``` to obtain the full reads, and use samtools ```get_reference_positions()``` to assign reads to their genome locations. 

I have been experimenting with both approaches, and would appreciate your input on the following:
* Using approach 1), I have noticed that the re-squiggling (has it been renamed to signal_refinement?) implemented in ```refine_signal_map[_core].py[x]```, treats insertions and deletions differently compared to the alignment stored in the bam files (in my case I used ```dorado```). Could you provide some insights into how indels of small size are handled during the DP approach? My ultimate goal is to aggregate the modified base information back to the bam files, which requires a one-to-one mapping and I currently am unclear of how to map a ```ReadRefReg``` object back to the corresponding ```AlignedSegment``` pysam object (or if that would even be possible).
* Using approach 2) I can obtain a one-to-one mapping of the signal to a particular base. However I am unsure if this approach would yield valid results as the alignment would have originated from sequence-to-sequence only and not signal-to-expected signal. 
* Directly related: for the same read, can we expect any difference in signal levels (when using the metrics API, e.g. mean and trimmed mean) between the ```ReadRefReg.norm_signal``` and ```io_read.norm_signal``` in the region corresponding to the defined ```RefRegion```? I have tried to understand this base on the source code, but these core methods are less documented compared to the higher level API. 
* In that same scope: does it make more sense to compare two samples using the reference-anchored approach, or would you recommend comparing on a basecall-anchored level and map afterwards? 

I would truly appreciate your insight into this. Thank you very much in advance and all the best.

**Comment by chilampoon** [comment]
Following up on this issue since I am doing similar things (on RNA) and still getting familiar with all these stuff - 

I noticed that some reads with multiple modifications don't align well with the reference sequence then they become unaligned reads.
I wonder if it is possible to assign the reads to the reference position based on the similarities of raw signals, like to do the 'alignment' by mapping raw signals of reads to raw signals of the reference (like [sigmap](https://github.com/haowenz/sigmap), but do it in remora and pod5? then compare signals on each signal-based aligned position.

For non DL methods, I am curious right now if it is still true that "the current mean intensity values and current intensity standard deviation were poor predictors of the modification status of the k-mer" as stated in epinano paper... also any pointers to papers/documentations for raw signal exploration would be appreciated. thank you very much

**Comment by marcus1487** [comment]
Lots of questions/clarifications, so I'll try to answer them specifically below.

> Let me begin by thanking you for providing us with the great API that was implemented in the later versions of remora. It feels like a big step forward in terms of providing consistent and easy access to signal based read analysis.

Glad you've found this interface useful! We are attempting to make this interface as useful as possible to thank you for the detailed questions here as it helps to tailor the documentation and future improvements to the interface!

> In order to build my canonical model for each nucleotide position of a reference genome, I need to assign the specific base of all reads covering the genomic position to that locus. I see two possible approaches here:
>
> Make use of the RefRegion class in conjunction with io.get_reads_reference_regions() and to extract all reads which cover a loci of interest and process the signal from there.
>
> Use io.get_reg_bam_reads() in conjunction with io.get_io_reads() to obtain the full reads, and use samtools get_reference_positions() to assign reads to their genome locations.

There are two concepts getting a bit mixed up here; this being basecall vs reference anchoring and signal map refinement (previously known as re-squiggling). Without considering signal map refinement, basecall anchored signal mapping simply uses the move table provided from the basecaller. A reference anchored signal map is then created by taking the move table along with the alignment to the reference sequence and composing these together to obtain a mapping from signal to reference bases ([implemented here](https://github.com/nanoporetech/remora/blob/master/src/remora/data_chunks.py#L54)). At reference deletions the signal positions are interpolated over the reference bases not mapped to basecalls. This gives a mapping from signal to reference bases. Thus for either basecall-anchored or reference anchored we end up with a sequence and a mapping from that sequence to the signal.

Signal mapping refinement is then performed from this starting point. Whether the sequence is basecalls or reference sequence is irrelevant to the signal mapping refinement step. In the signal mapping refinement there is a step to ensure that there is a wide enough band to allow the signal mapping refinement to search for the optimal path ([implemented here](https://github.com/nanoporetech/remora/blob/v2.1.3/src/remora/refine_signal_map_core.pyx#L31)). 

Note that this whole procedure is wrapped inside the io_read object, but happens each time this refined signal map is generated. Also note that this procedure is performed at the read level and not the region level, so if you need the metric from all reads covering a reference position, you'll have to perform signal mapping refinement over each read covering the position. This should be quite efficient, but repeating the same signal refinement (especially for very long reads) can become quite expensive. 

> I currently am unclear of how to map a ReadRefReg object back to the corresponding AlignedSegment pysam object (or if that would even be possible).

You can use the `make_sequence_coordinate_mapping` function to create a mapping from reference coordinates to basecall coordinates ([here](https://github.com/nanoporetech/remora/blob/v2.1.3/src/remora/data_chunks.py#L107)). In terms of how the a region of reference corresponds with the pysam object. 

> Using approach 2) I can obtain a one-to-one mapping of the signal to a particular base. However I am unsure if this approach would yield valid results as the alignment would have originated from sequence-to-sequence only and not signal-to-expected signal.

`get_io_reads` and `get_reg_bam_reads` do not perform signal mapping refinement. In order to compare any signal metrics across reads signal mapping refinement should be performed. 

>Directly related: for the same read, can we expect any difference in signal levels (when using the metrics API, e.g. mean and trimmed mean) between the ReadRefReg.norm_signal and io_read.norm_signal in the region corresponding to the defined RefRegion? I have tried to understand this base on the source code, but these core methods are less documented compared to the higher level API.

`ReadRefReg.norm_signal` is exactly the section of `io_read.norm_signal` mapped to the defined RefRegion defined by the current ref_to_signal mapping (whether signal mapping refinement was performed or not). 

> In that same scope: does it make more sense to compare two samples using the reference-anchored approach, or would you recommend comparing on a basecall-anchored level and map afterwards?

Two samples (or even any two reads) are really only comparable at the reference anchored level. It is possible in some cases that you might be able to make a direct comparison at the basecall level, but this would likely involve a mapping between the two. Signal metrics are also very unlikely to be useful without the signal mapping refinement step applied. 

This was as bit long winded, but do let me know if there are additional points that would be helpful to clarify and I do hope to get to some improved documentation on this interface in the medium term.

**Comment by marcus1487** [comment]
@chilampoon , I'm not familiar with sigmap, but if a mapping between signal and reference positions is produced then these could be input into a Remora io.read object. Then the remora downstream tools could be used. I would be quite surprised if sequence space mapping would fail, but signal space mapping would work though (especially in RNA). The Remora interface intentionally avoids adding any more than the essential bits of information to map signal to sequence to allow such exploration. If there are missing features once you start working more on this, do please raise them here. 

> For non DL methods, I am curious right now if it is still true that "the current mean intensity values and current intensity standard deviation were poor predictors of the modification status of the k-mer" as stated in epinano paper... also any pointers to papers/documentations for raw signal exploration would be appreciated. thank you very much

This is quite a complex question. The signal level predictors are very dependent upon the signal mapping refinement methods employed. It is certainly possible that signal level metrics can produce highly informative predictors, but we have found better performance internally by employing algorithms on raw/normalized signal values.

For pointers on raw signal analysis, the Remora documentation is likely the best resource. As noted it is a target to improve the documentation around this interface, but the questions addressed by raw signal are generally quite complex, so much of this is left to the individual researcher still.

**Comment by drivenbyentropy** [comment]
> Note that this whole procedure is wrapped inside the io_read object, but happens each time this refined signal map is generated. Also note that this procedure is performed at the read level and not the region level, so if you need the metric from all reads covering a reference position, you'll have to perform signal mapping refinement over each read covering the position. This should be quite efficient, but repeating the same signal refinement (especially for very long reads) can become quite expensive.

Just to make sure, you are referring to ```io_read.set_refine_signal_mapping(sig_map_refiner, ref_mapping=True)```, correct?

**Comment by marcus1487** [comment]
> Just to make sure, you are referring to io_read.set_refine_signal_mapping(sig_map_refiner, ref_mapping=True), correct?

Yes. That is correct.

> Does this then imply that in order to obtain the correct mapping, we would have to map all values to int, and ignore any values which have non-zero decimal values? I also believe that in order to obtain the correct reference base, we would have to offset i by the start position of the mapping, i.e. ref_seq[i+io_read.ref_reg.start].

This depends on what you mean by "correct". Over an indel there is not a "correct" reference base to which the basecall is to be mapped. Non-integer values are indeed the interpolation over an indel. You may decide that for particular applications truncating this value to give an integer reference position is the best choice. These values are left as float values here to enable the access to the closest signal point over an indel which may be within the signal assigned to one base and another in the basecalled sequence. The `ref_seq` in the `io.Read` object has already been extracted and converted to be the bit of sequence aligning to the read. If you were to grab the reference sequence from a FASTA then the `io_read.ref_reg.start` would have to be applied.

> Does this strategy sound reasonable to you?

I think this strategey is reasonable, but also sounds exactly like the logic used to extract the metrics in the `get_ref_reg_samples_metrics` method as shown in the [metrics API notebook](https://github.com/nanoporetech/remora/blob/v2.1.3/notebooks/metrics_api.ipynb). Is there something that you are extracting that is different from the output of you method? If you really wanted to be ultra-efficient you could save the reads which are re-squiggled for the next reference region, and add the new mapping reads, but this is certainly a larger engineering task.

**Comment by chilampoon** [comment]
Thank you both for these detailed questions and answers.

@marcus1487 I am confused by some definitions:
1. what is `stride` (the first value of the mv tag)? is it something related to the region of primer or barcode of a read (dacs before `num_trimmed`)?
2. I am trying to understand the indel interpolation in base to signal mappings:

for example I have a read where the first dozens of bases are like `8S13M1D7M2D1M`, then the first 8+21 queries of this alignment should map to the first 24 bases in the reference sequence after the reference start, then I looked at

```
io_read.seq[:8+21]
'AGGCAAGTACGTATCAGAGCGAACGAATT'

io_read.ref_seq[:24]
'ACGTATCGGAGCGAAACGAATCGT'
```

you can see

```
seq: AGGCAAGTACGTATCAGAGCG-AACGAAT--T
ref: SSSSSSSSACGTATCGGAGCGAAACGAATCGT
```

but the when I look at their coordinates:
```
io_read.ref_to_signal[:24+1]
array([ 225,  268,  314,  322,  327,  331,  391,  395,  634,  638,  685,
        696, 1102, 1106, 1110, 1114, 1118, 1127, 1131, 1135, 1140, 1247,
       1308, 1376, 1380], dtype=int32)
       
io_read.query_to_signal[8:8+21+1]
array([ 225,  255,  305,  410,  550,  620,  625,  650,  690,  705,  710,
        725,  775,  825, 1140, 1155, 1175, 1250, 1290, 1345, 2000, 2010])
```
do they look right? it seems to me that there are some shifts between these two series of coordinates, because after a few bases `ref_to_signal` coords go to 2000 something.

Then if I compare another read without any deletions like `8S24M`, is it to compare signals in `io_read.query_to_signal[8:8+24+1]` to the same region for the above read? Thank you.

**Comment by marcus1487** [comment]
> That is an interesting idea. Does the API provide functions for I/O operations on io_read objects? I am currently processing my genome in chunks on our HPC cluster to lower compute times to a manageable level, but at the expense that reads which overlap multiple regions are resquiggled multiple times.

@drivenbyentropy , The best pointer for the I/O operations would be in the infer code. You can see here where [MM/ML codes are swapped out for calls made by the remora infer command followed optionally by the reference anchored swapping of the basecalls for the reference seqeunce](https://github.com/nanoporetech/remora/blob/v2.1.3/src/remora/inference.py#L254) and are then [written to file here](https://github.com/nanoporetech/remora/blob/v2.1.3/src/remora/inference.py#L368).

**Comment by marcus1487** [comment]
> what is stride (the first value of the mv tag)? is it something related to the region of primer or barcode of a read (dacs before num_trimmed)?

Stride is a hyperparameter from the basecalling neural network describing the stride in the convolutional layers. This enforces that bases can only be output at this interval over the input signal. Thus to save space in the move tag only the positions every stride are stored. We are looking into adding more documentation on this to the basecaller repositories.

> do they look right? it seems to me that there are some shifts between these two series of coordinates, because after a few bases ref_to_signal coords go to 2000 something.

@chilampoon , for this one is it possible that you have run `set_refine_signal_mapping` on the read? This function will update either the query mapping OR the reference mapping depending on the specified `ref_mapping` argument to this function. If you are looking for a valid mapping between query to signal and reference to signal you should either apply this function to neither or both signal mappings. If you have not applied signal mapping refinement then this does look quite odd. Could you provide reproducible code to get these mapping results?

**Comment by chilampoon** [comment]
> @chilampoon , for this one is it possible that you have run `set_refine_signal_mapping` on the read? This function will update either the query mapping OR the reference mapping depending on the specified `ref_mapping` argument to this function. If you are looking for a valid mapping between query to signal and reference to signal you should either apply this function to neither or both signal mappings. If you have not applied signal mapping refinement then this does look quite odd. Could you provide reproducible code to get these mapping results?

Yeah I applied signal mapping refinement, here is the code

```
sig_map_refiner = refine_signal_map.SigMapRefiner(
    kmer_model_filename=level_table,
    scale_iters=0,
    do_fix_guage=True,
)
pod5_read = next(mod_pod5.reads(selection=[read_bam.query_name]))
io_read = io.Read.from_pod5_and_alignment(pod5_read, read_bam)
io_read.set_refine_signal_mapping(sig_map_refiner, ref_mapping=True)
```

I didn't run the refine_signal_mapping and got
```
io_read.ref_to_signal[:24+1]
array([ 225,  255,  305,  410,  550,  620,  625,  650,  690,  705,  710,
        725,  775,  800,  825, 1140, 1155, 1175, 1250, 1290, 1345, 1563,
       1781, 2000, 2080])
       
io_read.query_to_signal[8:8+21+1]
array([ 225,  255,  305,  410,  550,  620,  625,  650,  690,  705,  710,
        725,  775,  825, 1140, 1155, 1175, 1250, 1290, 1345, 2000, 2010])
```
Looks fine now, but if Im gonna compare signals of different reads on the same region, should I do the refinement?

**Comment by chilampoon** [comment]
I revisit previous comments, `io_read.query_to_signal` is the result of compute_ref_to_signal(), is the analysis order like, take dacs inside the reference coordinates for reads, then do set_refine_signal_mapping() on each, then compute metrics etc..? I probably need more time to look deeper :/

**Comment by drivenbyentropy** [comment]
> @drivenbyentropy , The best pointer for the I/O operations would be in the infer code. You can see here where [MM/ML codes are swapped out for calls made by the remora infer command followed optionally by the reference anchored swapping of the basecalls for the reference seqeunce](https://github.com/nanoporetech/remora/blob/v2.1.3/src/remora/inference.py#L254) and are then [written to file here](https://github.com/nanoporetech/remora/blob/v2.1.3/src/remora/inference.py#L368).

This is perfect, thank you. I had implemented a rudimentary version of this functionality for a different project where I was mapping ```tombo```s output back into the BAM files, but looking at your code references, I should be able to accomplish the same by using API calls only. @marcus1487 Thanks again for your detailed help in this, it is truly appreciated.
